source/Timer.cpp

```diff
@@ -0,0 +1,389 @@
+	// conditions is specified for the timer, since we are then necessarily outside the zone, and thus the
+	// clock is stopped.
+	reset |= (cond == Timer::ResetCondition::LEAVE_SYSTEM && (resetCondition == Timer::ResetCondition::PAUSE
+				|| resetCondition == Timer::ResetCondition::LEAVE_ZONE));
+	if(isActive && reset)
+	{
+		timeElapsed = 0;
+		// Perform the reset action, if there is one, assuming *either* it
+		// hasn't fired yet, *or* the timer is configured to fire it every reset.
+		if(repeatReset || !resetFired)
+		{
+			auto it = actions.find(TimerTrigger::RESET);
+			if(it != actions.end())
+				it->second.Do(player, ui, &mission);
+			resetFired = true;
+		}
+		isActive = false;
+	}
+}
+
+
+
+void Timer::Step(PlayerInfo &player, UI *ui, const Mission &mission)
+{
+	// Don't do any work for already-completed timers.
+	if(isComplete)
+		return;
+	const Ship *flagship = player.Flagship();
+	// Since we can only advance timers while flying, we can't do that if we have no flagship.
+	if(!flagship)
+		return;
+	// Now we go through all the possible reset/do-not-advance conditions for the timer.
+	// If and only if it passes all of them, the timer is marked as active and it advances.
+	// For each, if it fails, it calls ResetOn() with that reset condition.
+
+	// First we check to see if the player is in the specified system (or one of them).
+	if((system && flagship->GetSystem() != system) ||
+		(!systems.IsEmpty() && !systems.Matches(flagship->GetSystem())))
+	{
+		ResetOn(Timer::ResetCondition::LEAVE_SYSTEM, player, ui, mission);
+		return;
+	}
+
+	// Then we check if the timer requires the player to be idle (not turning, accelerating, or moving faster than
+	// the idle max speed) or peaceful (not firing any weapons).
+	if(requireIdle || requirePeaceful)
+	{
+		bool shipIdle = true;
+		if(requireIdle)
+			shipIdle = (!flagship->IsThrusting() && !flagship->IsSteering()
+						&& !flagship->IsReversing() && flagship->Velocity().LengthSquared() < idleMaxSpeed);
+		if(requirePeaceful)
+			for(const Hardpoint &weapon : flagship->Weapons())
+				shipIdle &= !weapon.WasFiring();
+		if(!shipIdle)
+		{
+			ResetOn(Timer::ResetCondition::PAUSE, player, ui, mission);
+			return;
+		}
+	}
+
+	// Then we check if the flagship is required to be uncloaked.
+	if(requireUncloaked)
+	{
+		double cloak = flagship->Cloaking();
+		if(cloak != 0.)
+		{
+			ResetOn(Timer::ResetCondition::PAUSE, player, ui, mission);
+			return;
+		}
+	}
+
+	// Finally, we check if the flagship is required to be close to a particular stellar object.
+	if(proximity > 0.)
+	{
+		bool inProximity = false;
+		if(proximityCache.size() > 0)
+			for(const StellarObject *proximityObject : proximityCache)
+			{
+				double dist = flagship->Position().Distance(proximityObject->Position());
+				if((closeTo && dist <= proximity) || (!closeTo && dist >= proximity))
+				{
+					inProximity = true;
+					break;
+				}
+			}
```

Comment:
if proximityCache is empty, the loop doesn't do anything anyway; I'd also use a ternary for the proximity conditions
```suggestion
		for(const StellarObject *proximityObject : proximityCache)
		{
			double dist = flagship->Position().Distance(proximityObject->Position());
			if(closeTo ? dist <= proximity : dist >= proximity)
			{
				inProximity = true;
				break;
			}
		}
```

Reply:
If you really want to simplify the conditional just do `closeto != dist >= proximity`
Even though that would ignore the single case of == but that's not important for double values

Reply:
> if proximityCache is empty, the loop doesn't do anything anyway; I'd also use a ternary for the proximity conditions

While this is more concise code, it's also much harder to parse on first look, IMO. I think it would be much clearer to just assign the value of the ternary to `inProximity`, then test `inProximity` and break if true.